Add state preparation

johannesjmeyer opened the pull request:
**Context:**
To support state preparations on hardware, a conversion from a state to a series of gates is necessary. This is achieved i.e. using the Möttönen state preparation.

**Description of the Change:**
This pull requests adds two new methods to the PennyLane template section:
1. `BasisStatePreparation`: simply apply `PauliX` gates to all wires that have a 1
2. `MottonenStatePreparation`: use the Möttönen state preparation to construct an arbitrary state

Both functions also work on subsets of wires.

**Benefits:**
Plugins can now easily implement state preparations on hardware by simply resorting to the above templates.

**Possible Drawbacks:**
None.

**Related GitHub Issues:**
#273

josh146 commented:
Thanks for this @johannesjmeyer. A `BasisStatePreparation` template is actually really useful as well!

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,103 @@
+# limitations under the License.
+r"""
+State Preparations
+==========
```
```suggestion
==================
```

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,103 @@
+
+    Args:
+        basis_state (array): Input array of shape ``(N,)``, where N is the number of qubits,
+            with :math:`N\leq n`
```
what is `n` in this case?

johannesjmeyer replied:
Number of wires. I copied this from the other templates. I'll change it in both places.

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,103 @@
+
+    if not isinstance(wires, Iterable):
+        raise ValueError(
+            "Wires needs to be a list of wires that the embedding uses; got {}.".format(wires)
```
This sounds a bit awkward, how about
```suggestion
            "Wires must be passed as a list of integers; got {}.".format(wires)
```

josh146 replied:
(Does this require that `wires` be an iterable even if the state is a one-qubit state?)

johannesjmeyer replied:
That's also something I copied from the other template, didn't actually think too much about it.

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,103 @@
+            qml.PauliX(wire)
+
+
+
```
```suggestion
```

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,103 @@
+
+    Args:
+        state_vector (array): Input array of shape ``(2^N,)``, where N is the number of qubits,
+            with :math:`N\leq n`
```
same question here re: `n`

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,103 @@
+            "State vector probabilities have to sum up to 1.0, got {}".format(probability_sum)
+        )
+        
+    # TODO: Implement the state preparation
```
Oh, before I go any further, is this WIP still? 🙂 

If so, can you mark it with the WIP label.

johannesjmeyer commented:
This is now ready for review. 

One thing that confused me was that the state preparation will always yield states that have a very high fidelity compared to the input state (this is also what I am testing right now), but the state vectors will look rather different. Is this a behaviour that was anticipated?

josh146 commented on pennylane/templates/embeddings.py:
```diff
@@ -69,16 +70,16 @@ def AngleEmbedding(features, wires, rotation='X'):
 
     if not isinstance(wires, Iterable):
-        raise ValueError("Wires needs to be a list of wires that the embedding uses; got {}.".format(wires))
+        raise ValueError("Wires must be passed as a list of integers; got {}.".format(wires))
```
Reads much better now!

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+
+
+def _matrix_M_entry(row, col):
+    """The matrix entry for the angle computation.
```
What does this function do/what is an angle computation? Might be worth expanding this docstring.

johannesjmeyer replied:
I had no idea myself, but I looked it up in the paper and its just the matrix mapping from alpha to theta.

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+
+        b_and_g = b_and_g >> 1
+
+    return (-1) ** sum_of_ones
```
This function is very hard to understand 😆 Would be good to add comments explaining the process.

johannesjmeyer replied:
Basically it implements the construction of M from the paper, I just copied it from Carsten Blank and I simply can't explain it :sweat_smile:

antalszava replied:
Glanced into the paper a bit and it seems that matrix M, whose entries this function builds up, is responsible for the mapping between the angles inputted to the uniformly controlled rotation (``thetas``) and the angles of the single qubit rotations (``alphas``). This is the mapping that yields the new parameters when the uniformly controlled rotation is decomposed into single qubit rotations and CNOTs.

In [their previous paper] it is stated: "The position of the control node in the *lth* CNOT gate is set to
match the position where the *l<sup>th</sup>* and *(l + 1)<sup>th</sup>* bit strings *g<sub>l−1</sub>* and *g<sub>l</sub>* of the binary reflected Gray code [18] differ.

So eventually these positions are computed here for each of the CNOTs and stored in the M matrix.

[their previous paper]: https://arxiv.org/abs/quant-ph/0404089

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+        np.unravel_index(range(2 ** n), [2] * n, order="C"), [2] * n, order="F"
+    )
+
+    state_vector = np.array(state_vector)[reorder_indices, np.newaxis]
```
A cleaner way of re-ordering the qubit order is to take the transpose:
```python
state_vector = state_vector.reshape([2] * n).T.flatten()
```

johannesjmeyer replied:
I actually wanted to know how I can shorten this. Thanks!

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+    #     print(i, ": control = ", control_indices[i])
+
+    #     qml.CNOT(wires=[control_wires[control_indices[i]], target_wire])
+    #     gate(theta[i, 0], wires=[target_wire])
```
Can the above be deleted?

josh146 commented on tests/test_templates_state_preparations.py:
```diff
@@ -0,0 +1,218 @@
+        #     #print(bin_maps)
+        #     for i in range(16):
+        #         print("{} -> {}, {} -> {}".format(idx_maps[i][0], idx_maps[i][1], bin_maps[i][0], bin_maps[i][1]))
+        #     raise Exception("X")
```
How come this test is commented out?

johannesjmeyer replied:
It was the test I used to get the code working, I thought I might still need it

josh146 commented on tests/test_templates_state_preparations.py:
```diff
@@ -0,0 +1,218 @@
+        ([0], 0, "Wires must be passed as a list of integers"),
+        ([1/2, 1/2], [0], "State vector probabilities have to sum up to 1.0"),
+        ([2/3, 0, 2j/3, -2/3], [0, 1], "State vector probabilities have to sum up to 1.0"),
+    ])
```
This is a matter of style, but I find this very hard to read. Ideally, this should likely be separate atomic tests (e.g., `test_invalid_wires_type`, `test_invalid_state` etc.)

johannesjmeyer replied:
Will do!

josh146 commented on tests/test_templates_state_preparations.py:
```diff
@@ -0,0 +1,218 @@
+
+        # print("state: ", state)
+        # print("target: ", target_state)
+        # print("fidelity: ", fidelity)
```
These should be deleted

josh146 commented on tests/test_templates_state_preparations.py:
```diff
@@ -0,0 +1,218 @@
+        ([1/2, 0, 0, 0, 1j/2, 0, 1j/math.sqrt(2), 0], [0, 1, 2], [1/2, 0, 0, 0, 1j/2, 0, 1j/math.sqrt(2), 0]),
+        ([1/2, 0, 1j/2, 1j/math.sqrt(2)], [0, 1], [1/2, 0, 0, 0, 1j/2, 0, 1j/math.sqrt(2), 0]),
+    ])
+    # fmt: on
```
As mentioned below, I find this very difficult to read due to the indentation. Can we move this to the top of the file as a `TEST_STATE_DATA` variable? Then the test becomes the much more readable
```python
    @pytest.mark.parametrize("state_vector,wires,target_state", TEST_STATE_DATA)
    def test_state_preparation(self, tol, qubit_device_3_wires, state_vector, wires, target_state):
```
and it could avoid some of the black formatting issues.

johannesjmeyer replied:
I think having the test data close to the tests is better. Otherwise you have the same blob of data somewhere else and you have to scroll. 

And i purposedly turned black off on the arguments because if black formats them it gets really really long. I think how it is is a good middle way.

josh146 replied:
In general I agree, but in this case the test data is quite large, so it impacts readability

antalszava replied:
How about having line breaks in the longer lists and then having an empty line between lists?
```python
(
            [-0.17133152-0.18777771j, 0.00240643-0.40704011j, 0.18684538-0.36315606j,
             -0.07096948+0.104501j, 0.30357755-0.23831927j, -0.38735106+0.36075556j,
             0.12351096-0.0539908j, 0.27942828-0.24810483j],
            
            [0, 1, 2],
            
            [-0.17133152-0.18777771j, 0.00240643-0.40704011j, 0.18684538-0.36315606j,
             -0.07096948+0.104501j, 0.30357755-0.23831927j, -0.38735106+0.36075556j,
             0.12351096-0.0539908j, 0.27942828-0.24810483j],
        ),
```
It could boost readability and would be close by the test at the same time.

johannesjmeyer commented:
I applied the review suggestions and implemented the graycode myself (which was actually a one-liner) and removed the sympy dependence.

Concerning the points you raised:

> It might be worth adding a default.qubit integration test to ensure that the probabilities are identical.

Yes, this is a very good idea. I'll do that!

> A separate thought I had: since the QNode knows what operations the device supports, it could seamlessly use the template in the background if qml.BasisState or qml.QubitStateVector is not supported. This allows plugins to continue to use their own decompositions if available, but allows those without to still support it.

I'd leave this to the plugins actually. If they want to support `qml.QubitStateVector` and `qml.BasisState` then they can use the templates. But maybe this is not desirable, so I would not make it an automatism.

johannesjmeyer commented:
I added a probability vector test.

antalszava commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+
+    Args:
+        basis_state (array): Input array of shape ``(N,)``, where N is the number of wires
+            the state preparation acts on. ``N`` must be smaller or equal to the total
```
Maybe it would nice to simply have ` ``N`` must be equal ` here in the docstring, as the statement on line 69 checks for that.

antalszava commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+    return theta
+
+
+def _uniform_rotation_dg(gate, alpha, control_wires, target_wire):
```
Writing out the word `dagger` in the function name could increase readability of the code (for each of the later functions as well).

antalszava commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+        qml.CNOT(wires=[control_wires[control_index], target_wire])
+
+
+def _unirz_dg(alpha, control_wires, target_wire):
```
Elaborating/splitting the `unirz` part of the function name such as `uni_rz`, `uni_zrot` or `uniformly_controlled_zrot` could also be nice for the reader

antalszava commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,315 @@
+    Args:
+        rank (int): rank of the Gray code (i.e. number of bits)
+    """
+    return ["{0:0{1}b}".format(i ^ (i >> 1), rank) for i in range(0, 1 << rank)]
```
Turns out, that this can be fastened by using a resursive approach (the one described here:
https://stackoverflow.com/questions/38738835/generating-gray-codes)

It really performs considerably better for larger systems (on my machine starting from around 20 qubits).

Simple script that can be nice to see how they perform:
```python
import time

def so_gray_code(n):
    def gray_code_recurse (g,n):
        k=len(g)
        if n<=0:
            return

        else:
            for i in range (k-1,-1,-1):
                char='1'+g[i]
                g.append(char)
            for i in range (k-1,-1,-1):
                g[i]='0'+g[i]

            gray_code_recurse (g,n-1)

    g=['0','1']
    gray_code_recurse(g,n-1)
    return g

def gray_code(rank):
    """Generates the Gray code of given rank.

    Args:
        rank (int): rank of the Gray code (i.e. number of bits)
    """
    return ["{0:0{1}b}".format(i ^ (i >> 1), rank) for i in range(0, 1 << rank)]

def func(n):
    start = time.time()
    a = gray_code(n)
    print(time.time()-start)
    b = so_gray_code(n)
    print(time.time()-start)
```
On my machine for ``rank==24`` ``gray_code`` ran for 7.7 seconds, ``so_gray_code`` for 3 seconds, at ``rank==20`` the difference is 0.2 seconds (pretty negligable).

johannesjmeyer replied:
Thanks a lot for this!

I also thought that it might be advantageous to use lazy evaluation here, because the gray code has an exponential number of elements. For many qubits this will likely create a huge memory consumption to create the whole gray code at once.

johannesjmeyer replied:
But I think we postpone this until people run out of memory. You will always have large memory consumption because you also use exponentially many gates, so I think the Gray code is not the most memory-demanding part of this anyways.

johannesjmeyer commented:
Thanks @antalszava, I applied your suggestions!

josh146 commented on doc/references.bib:
```diff
@@ -0,0 +1,453 @@
+  AUTHOR = {Schuld, Maria and Petruccione, Francesco},
+  YEAR = {2018}, 
+  PUBLISHER = {Springer},
+}
```
In the QML/PennyLane repository split, PennyLane lost the bib file :) The bib file is now part of the QML repo, since it hosts the background information/theory.

As a result, docstrings in the PennyLane docs now simply have a handwritten reference/link to the arXiv.

So best to remove this file, and remove any references that were added.

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,330 @@
+    that is uniformly controlled by the control qubits.
+
+    Args:
+        gate (qml.Operation): gate to be applied, needs to have exactly
```
```suggestion
        gate (~.Operation): gate to be applied, needs to have exactly
```

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,330 @@
+        k (int): current qubit
+
+    Returns:
+        a sparse vector representing :math:`\alpha^z_k`
```
These should have a return type
```suggestion
        scipy.sparse.dok_matrix[np.float64]: a sparse vector representing :math:`\alpha^z_k`
```

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,330 @@
+        k (int): current qubit
+
+    Returns:
+        a sparse vector representing :math:`\alpha^y_k`
```
Missing a return type

josh146 commented on pennylane/templates/state_preparations.py:
```diff
@@ -0,0 +1,330 @@
+def MottonenStatePreparation(state_vector, wires):
+    r"""
+    Prepares an arbitrary state on the given wires using a decomposition into gates developed
+    by Möttönen et al. :cite:`mottonen2005transformation`. The state is prepared via a sequence
```
A line break will allow the summary to be properly rendered by autosummary:
```suggestion
    by Möttönen et al. :cite:`mottonen2005transformation`.
    
    The state is prepared via a sequence
```
